### ISSUE TITLE
Changed behavior of how InputText handle repeated calls to SetKeyboardFocusHere

### ISSUE THREAD

**Issue by corporateshark** [ocornut/imgui#4682]
**Version/Branch of Dear ImGui:**

Version: 1.86 WIP
Branch: docking (bac748fa95ac003c7b354139980f8b4b7f6ac5da)

**Back-end/Renderer/Compiler/OS**

Back-ends: custom input back-end + custom Vulkan renderer.
Compiler: VS 2019
Operating System: Windows 10

**My Issue/Question:**

My input back-end is handling all the keys and sending input to ImGui using `ImGuiIO::AddInputCharacter()` the following way:

```
const ImWchar ch = Utils::TranslateKey(key, Env->GUI->GetDesktop()->IsKeyPressed(LK_SHIFT));
io.AddInputCharacter(ch);
```

After upgrading to 1.86 WIP, my abovementioned code stopped working with `ImGui::InputTextEx()`. `ImGui::InputTextEx()` simply does not see any input. 

I traced ImGui and found this new code fragment inside `bool ImGui::InputTextEx()`.
```
        if (io.InputQueueCharacters.Size > 0)
        {
            if (!ignore_char_inputs && !is_readonly && !input_requested_by_nav)
                for (int n = 0; n < io.InputQueueCharacters.Size; n++)
                {
                    // Insert character if they pass filtering
                    unsigned int c = (unsigned int)io.InputQueueCharacters[n];
                    if (c == '\t' && io.KeyShift)
                        continue;
                    if (InputTextFilterCharacter(&c, flags, callback, callback_user_data, ImGuiInputSource_Keyboard))
                        state->OnKeyPressed((int)c);
                }

            // Consume characters
            io.InputQueueCharacters.resize(0);
        }
```        
Here the newly introduced variable `input_requested_by_nav` is always `true` in my case, which results in skipping all the input characters.

The last working commit on `docking` is 835a5344b01e79aca479e95fa9d36fb5fdef3d14.

Is it a bug or an API misuse from my side?

**Comment by ocornut** [comment]
Can you debug why input_requested_by_nav is true? That should happen when eg pressing Enter or triangle on a Gamepad.

**Comment by corporateshark** [comment]
I'm taking `docking` bac748fa95ac003c7b354139980f8b4b7f6ac5da and splitting `input_requested_by_nav` evaluation:

```
const bool input_requested_by_nav = (g.ActiveId != id) && ((g.NavActivateInputId == id) || (g.NavActivateId == id && g.NavInputSource == ImGuiInputSource_Keyboard));
```

Into 3 variables like this:

```
const bool isActiveId = (g.ActiveId != id); // true
const bool isActiveInputId = (g.NavActivateInputId == id); // true
const bool isNavActivatedId = (g.NavActivateId == id && g.NavInputSource == ImGuiInputSource_Keyboard); // false

const bool input_requested_by_nav = isActiveId && (isActiveInputId || isNavActivatedId); // true
```

**Comment by ocornut** [comment]
Then you want to check why g. NavActivateInputId is set in NavUpdate() at beginning of frame. At you calling SetKeyboardFocusHere() every frame by any chance?

**Comment by corporateshark** [comment]
> At you calling SetKeyboardFocusHere() every frame by any chance?
Yes... Here's the code which invokes this `ImGui::InputText()`.

```
const ImGuiInputTextFlags flags =
   ImGuiInputTextFlags_EnterReturnsTrue |
   ImGuiInputTextFlags_CallbackCompletion |
   ImGuiInputTextFlags_CallbackHistory |
   ImGuiInputTextFlags_CallbackEdit |
   ImGuiInputTextFlags_CallbackCharFilter;
if (ImGui::InputText("##ConsoleInput", FTextBuffer, IM_ARRAYSIZE(FTextBuffer), flags, [](ImGuiInputTextCallbackData* data) -> int { return ((clGUIStdConsole*)(data->UserData))->TextEditCallback(data); }, this))
{
	ExecuteCommand();
}
ImGui::SetKeyboardFocusHere(-1); // auto focus the previous widget
```

**Comment by corporateshark** [comment]
Replacing it with this code solved the problem.

```
if (FReclaimFocus) {
	ImGui::SetKeyboardFocusHere(-1); // auto focus the previous widget
	FReclaimFocus = false;
}
```

**Comment by ocornut** [comment]
I’ll note as a regression that refocusing every frame breaks this. Will fix.

**Comment by corporateshark** [comment]
Thank you so much!

**Comment by ocornut** [comment]
Pushed a fix for this ad7e062b30c93935429889c1db99a4287556cc12
You can easily cherry-pick it in your branch.
